Pal/security/linux/graphene-sandbox.c

```diff
@@ -0,0 +1,1037 @@
+{
+#define ADDR_STR_MAX	128
+
+	char str[ADDR_STR_MAX];
+	int len = 0;
+
+	if (n->flags & ADDR_ANY) {
+		str[len++] = 'A';
+		str[len++] = 'N';
+		str[len++] = 'Y';
+		str[len++] = ':';
+	} else {
+		switch(n->family) {
+		case AF_INET: {
+			u8 *ip = (u8 *) &n->addr.addr.sin_addr.s_addr;
+			len += snprintf(str + len,
+					ADDR_STR_MAX - len,
+					"%u.%u.%u.%u:",
+					ip[0], ip[1], ip[2], ip[3]);
+			}
+			break;
+		case AF_INET6: {
+			u16 *ip = (u16 *) &n->addr.addr.sin6_addr.s6_addr;
+			len += snprintf(str + len,
+					ADDR_STR_MAX - len,
+					"[%u:%u:%u:%u:%u:%u:%u:%u]:",
+					ip[0], ip[1], ip[2], ip[3],
+					ip[4], ip[5], ip[6], ip[7]);
+			}
+			break;
+		}
+	}
+
+	if (n->flags & PORT_ANY) {
+		str[len++] = 'A';
+		str[len++] = 'N';
+		str[len++] = 'Y';
+	} else {
+		if (n->addr.port_begin == n->addr.port_end)
+			len += snprintf(str + len, ADDR_STR_MAX - len,
+					"%u", n->addr.port_begin);
+		else
+			len += snprintf(str + len, ADDR_STR_MAX - len,
+					"%u-%u",
+					n->addr.port_begin, n->addr.port_end);
+	}
+
+	BUG_ON(len >= ADDR_STR_MAX);
+	str[len] = 0;
+	printk(fmt, current->pid, str);
+}
+#else
+# define print_net_rule(...) do {} while (0)
+#endif
+
+static int set_net_rule(struct graphene_net_rule *nr, struct graphene_info *gi,
+			bool bind)
+{
+	struct graphene_net *n;
+
+	if (nr->family != AF_INET && nr->family != AF_INET6)
+		return -EINVAL;
+
+	n = kmalloc(sizeof(struct graphene_net), GFP_KERNEL);
+	if (!n)
+		return -ENOMEM;
+
+	n->family  = nr->family;
+	n->flags   = 0;
+	n->addr    = nr->addr;
+
+	switch(n->family) {
+	case AF_INET:
+		if (!n->addr.addr.sin_addr.s_addr)
+			n->flags |= ADDR_ANY;
+		break;
+	case AF_INET6:
+		if (!memcmp(&n->addr.addr.sin6_addr.s6_addr, &in6addr_any, 16))
+			n->flags |= ADDR_ANY;
+		break;
+	}
+
+	if (n->addr.port_begin == 0 && n->addr.port_end == 65535)
+		n->flags |= PORT_ANY;
+
+	INIT_LIST_HEAD(&n->list);
+	if (bind) {
+		list_add_tail(&n->list, &gi->gi_binds);
+		print_net_rule(KERN_INFO "Graphene: PID %d NET BIND %s\n", n);
+	} else {
+		list_add_tail(&n->list, &gi->gi_peers);
+		print_net_rule(KERN_INFO "Graphene: PID %d NET PEER %s\n", n);
+	}
+	return 0;
+}
+
+static int update_sandbox(struct file *file, struct graphene_info *new);
+
+int set_sandbox(struct file *file,
+		const struct graphene_policies __user *gpolicies)
+{
+	const struct graphene_user_policy __user *policies;
+	int npolicies;
+	struct graphene_info *gi;
+	struct graphene_user_policy ptmp;
+	struct graphene_path *p;
+	struct filename *tmp = NULL, *new;
+	char *kname;
+	int rv, i;
+
+	policies = gpolicies->policies;
+	rv = copy_from_user(&npolicies, &gpolicies->npolicies, sizeof(int));
+	if (rv)
+		return -EFAULT;
+
+	if (npolicies && !policies)
+		return -EINVAL;
+
+	gi = kmalloc(sizeof(struct graphene_info), GFP_KERNEL);
+	if (!gi)
+		return -ENOMEM;
+
+	tmp = __getname();
+	rv = -ENOMEM;
+	if (unlikely(!tmp))
+		goto out;
+
+	kname = (char *) tmp;
+	memset(gi, 0, sizeof(struct graphene_info));
```

Comment:
`memset(gi, 0, sizeof(*gi));` is shorter and safer - if during some refactoring `gi` change its type, this code will silently corrupt memory. It'd be nice to fix similar occurences in other places also.